sig-storage: Mutable CSI Node Allocatable Blog Graduates to Beta (#51536)
kubernetes/website: content/en/blog/_posts/2025-07-07-mutable-csi-node-allocatable.md

---
layout: blog
title: "Kubernetes v1.34: Mutable CSI Node Allocatable Graduates to Beta"

> date: 2025-07-07

Reviewer: ```suggestion
date: 2025-XX-XX
```

nit

slug: kubernetes-v1-34-mutable-csi-node-allocatable-count
draft: true
author: Eddie Torres (Amazon Web Services)
---

> The [functionality for CSI drivers to update information about attachable volume count on the nodes](https://github.com/kubernetes/enhancements/issues/4876), first introduced as Alpha in Kubernetes v1.33, has graduated to **Beta** in the Kubernetes v1.34 release! This marks a significant milestone in enhancing the accuracy of stateful pod scheduling by reducing failures due to outdated attachable volume capacity information.

Reviewer: ```suggestion
The [functionality for CSI drivers to update information about attachable volume count on the nodes](https://kep.k8s.io/4876), first introduced as Alpha in Kubernetes v1.33, has graduated to **Beta** in the Kubernetes v1.34 release! This marks a significant milestone in enhancing the accuracy of stateful pod scheduling by reducing failures due to outdated attachable volume capacity information.
```

## Background

> Traditionally, Kubernetes CSI drivers report a static maximum volume attachment limit when initializing. However, actual attachment capacities can change during a node's lifecycle for various reasons, such as:

Reviewer: ```suggestion
Traditionally, Kubernetes [CSI drivers](https://kubernetes-csi.github.io/docs/introduction.html) report a static maximum volume attachment limit when initializing. However, actual attachment capacities can change during a node's lifecycle for various reasons, such as:
```

## Dynamically adapting CSI volume limits

With this new feature, Kubernetes enables CSI drivers to dynamically adjust and report node attachment capacities at runtime. This ensures that the scheduler, as well as other components relying on this information, have the most accurate, up-to-date view of node capacity.

### How it works

Kubernetes supports two mechanisms for updating the reported node volume limits:

- **Periodic Updates:** CSI drivers specify an interval to periodically refresh the node's allocatable capacity.
- **Reactive Updates:** An immediate update triggered when a volume attachment fails due to exhausted resources (`ResourceExhausted` error).

### Enabling the feature

To use this beta feature, the `MutableCSINodeAllocatableCount` feature gate must be enabled in these components:

- `kube-apiserver`
- `kubelet`

### Example CSI driver configuration

Below is an example of configuring a CSI driver to enable periodic updates every 60 seconds:

```
apiVersion: storage.k8s.io/v1
kind: CSIDriver
metadata:
  name: example.csi.k8s.io
spec:
  nodeAllocatableUpdatePeriodSeconds: 60
```

> This configuration directs Kubelet to periodically call the CSI driver's `NodeGetInfo` method every 60 seconds, updating the node’s allocatable volume count. Kubernetes enforces a minimum update interval of 10 seconds to balance accuracy and resource usage.

Reviewer: ```suggestion
This configuration directs kubelet to periodically call the CSI driver's `NodeGetInfo` method every 60 seconds, updating the node’s allocatable volume count. Kubernetes enforces a minimum update interval of 10 seconds to balance accuracy and resource usage.
```

---
layout: blog
title: "Kubernetes v1.34: Mutable CSI Node Allocatable Graduates to Beta"
date: 2025-07-07

> slug: kubernetes-1-34-mutable-csi-node-allocatable-count

Reviewer 2: ```suggestion
slug: kubernetes-v1-34-mutable-csi-node-allocatable-count
```

draft: true
author: Eddie Torres (Amazon Web Services)
---

> The *mutable CSI node allocatable count* feature, first introduced as Alpha in Kubernetes v1.33, has graduated to **Beta** and is enabled by default in the Kubernetes v1.34 release! This marks a significant milestone in enhancing the accuracy of stateful pod scheduling by reducing failures due to outdated volume capacity information.

Author: Thank you, done.

> The *mutable CSI node allocatable count* feature, first introduced as Alpha in Kubernetes v1.33, has graduated to **Beta** and will be enabled by default in the Kubernetes v1.34 release! This marks a significant milestone in enhancing the accuracy of stateful pod scheduling by reducing failures due to outdated volume capacity information.

Reviewer 3: Hey, when this publishes, the release will be in the past. You should word it accordingly.

- Multi-driver scenarios, where one CSI driver’s operations affect available capacity reported by another.

Static reporting can cause Kubernetes to schedule pods onto nodes that appear to have capacity but don't, leading to pods stuck in a `ContainerCreating` state.

## Dynamically adapting CSI volume limits

With the new feature gate `MutableCSINodeAllocatableCount`, Kubernetes enables CSI drivers to dynamically adjust and report node attachment capacities at runtime. This ensures that the scheduler has the most accurate, up-to-date view of node capacity.

### How it works

When this feature is enabled, Kubernetes supports two mechanisms for updating the reported node volume limits:

- **Periodic Updates:** CSI drivers specify an interval to periodically refresh the node's allocatable capacity.
- **Reactive Updates:** An immediate update triggered when a volume attachment fails due to exhausted resources (`ResourceExhausted` error).

### Enabling the feature

To use this beta feature, the `MutableCSINodeAllocatableCount` feature gate must be enabled in these components:

- `kube-apiserver`
- `kubelet`

> The feature gate is enabled by default in Kubernetes 1.34+.

Reviewer 4: Roger, that makes more sense.
